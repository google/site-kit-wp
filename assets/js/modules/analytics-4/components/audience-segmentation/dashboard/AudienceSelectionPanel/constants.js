/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
