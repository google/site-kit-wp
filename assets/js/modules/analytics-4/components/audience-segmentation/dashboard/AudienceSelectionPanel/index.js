/**
 * Audience Selection Panel
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { AUDIENCE_SELECTION_PANEL_OPENED_KEY } from './constants';
import { CORE_UI } from '../../../../../../googlesitekit/datastore/ui/constants';
import InViewProvider from '../../../../../../components/InViewProvider';
import Panel from './Panel';

export default function AudienceSelectionPanel() {
	const isOpen = useSelect( ( select ) =>
		select( CORE_UI ).getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
	);

	return (
		<InViewProvider
			value={ {
				key: 'AudienceSelectionPanel',
				value: !! isOpen,
			} }
		>
			<Panel />
		</InViewProvider>
	);
}
