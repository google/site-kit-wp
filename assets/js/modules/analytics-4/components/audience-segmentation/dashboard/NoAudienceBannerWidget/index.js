/**
 * NoAudienceBannerWidget component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import whenActive from '../../../../../../util/when-active';
import { MODULES_ANALYTICS_4 } from '../../../../datastore/constants';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import NoAudienceBanner from './NoAudienceBanner';
import withIntersectionObserver from '../../../../../../util/withIntersectionObserver';
import { trackEvent } from '../../../../../../util';
import useViewContext from '../../../../../../hooks/useViewContext';

const NoAudienceBannerWithIntersectionObserver =
	withIntersectionObserver( NoAudienceBanner );

function NoAudienceBannerWidget( { Widget, WidgetNull } ) {
	const viewContext = useViewContext();

	const availableAudiences = useSelect( ( select ) => {
		const audiences = select( MODULES_ANALYTICS_4 ).getAvailableAudiences();
		return audiences?.map( ( audience ) => audience.name );
	} );

	const configuredAudiences = useSelect( ( select ) =>
		select( CORE_USER ).getConfiguredAudiences()
	);

	const didSetAudiences = useSelect( ( select ) =>
		select( CORE_USER ).didSetAudiences()
	);

	const hasNoMatchingAudience = configuredAudiences?.every(
		( audience ) =>
			Array.isArray( availableAudiences ) &&
			! availableAudiences.includes( audience )
	);
	if (
		!! configuredAudiences &&
		( configuredAudiences?.length === 0 || hasNoMatchingAudience )
	) {
		return (
			<Widget noPadding>
				<NoAudienceBannerWithIntersectionObserver
					onInView={ () => {
						trackEvent(
							`${ viewContext }_audiences-no-audiences`,
							'view_banner',
							didSetAudiences
								? 'no-longer-available'
								: 'none-selected'
						);
					} }
				/>
			</Widget>
		);
	}

	return <WidgetNull />;
}

NoAudienceBannerWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	WidgetNull: PropTypes.elementType.isRequired,
};

export default whenActive( { moduleName: 'analytics-4' } )(
	NoAudienceBannerWidget
);
