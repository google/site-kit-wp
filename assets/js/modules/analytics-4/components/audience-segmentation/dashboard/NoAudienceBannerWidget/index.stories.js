/**
 * NoAudienceBannerWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies.
 */
import NoAudienceBannerWidget from '.';
import WithRegistrySetup from '../../../../../../../../tests/js/WithRegistrySetup';
import {
	provideModuleRegistrations,
	provideModules,
} from '../../../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '../../../../../../googlesitekit/widgets/util';
import { MODULES_ANALYTICS_4 } from '../../../../datastore/constants';
import { availableAudiences } from '../../../../datastore/__fixtures__';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import { Provider as ViewContextProvider } from '../../../../../../components/Root/ViewContextContext';
import {
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
} from '../../../../../../googlesitekit/constants';

const WidgetWithComponentProps = withWidgetComponentProps(
	'analyticsNoAudienceBanner'
)( NoAudienceBannerWidget );

function Template() {
	return <WidgetWithComponentProps />;
}

export const AuthenticatedUserNeverSetup = Template.bind( {} );
AuthenticatedUserNeverSetup.storyName =
	'Authenticated user, never populated their audience selection';
AuthenticatedUserNeverSetup.scenario = {};
AuthenticatedUserNeverSetup.args = {
	setupRegistry: ( registry ) => {
		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [],
			didSetAudiences: false,
		} );
	},
};

export const AuthenticatedUserHasSetup = Template.bind( {} );
AuthenticatedUserHasSetup.storyName =
	'Authenticated user, previously populated their audience selection';
AuthenticatedUserHasSetup.scenario = {};
AuthenticatedUserHasSetup.args = {
	setupRegistry: ( registry ) => {
		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [],
			didSetAudiences: true,
		} );
	},
};

export const ViewOnlyUserNeverSetup = Template.bind( {} );
ViewOnlyUserNeverSetup.storyName =
	'View-only user, never populated their audience selection';
ViewOnlyUserNeverSetup.scenario = {};
ViewOnlyUserNeverSetup.args = {
	setupRegistry: ( registry ) => {
		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [],
			didSetAudiences: false,
		} );
	},
	isViewOnly: true,
};

export const ViewOnlyUserHasSetup = Template.bind( {} );
ViewOnlyUserHasSetup.storyName =
	'View-only user, previously populated their audience selection';
ViewOnlyUserHasSetup.scenario = {};
ViewOnlyUserHasSetup.args = {
	setupRegistry: ( registry ) => {
		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [],
			didSetAudiences: true,
		} );
	},
	isViewOnly: true,
};

export default {
	title: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/NoAudienceBannerWidget',
	decorators: [
		( Story, { args } ) => {
			const setupRegistry = async ( registry ) => {
				provideModules( registry, [
					{
						active: true,
						connected: true,
						slug: 'analytics-4',
					},
				] );
				provideModuleRegistrations( registry );

				registry
					.dispatch( MODULES_ANALYTICS_4 )
					.setAvailableAudiences( availableAudiences );

				await args.setupRegistry( registry );
			};

			const viewContext = args.isViewOnly
				? VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY
				: VIEW_CONTEXT_MAIN_DASHBOARD;

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<ViewContextProvider value={ viewContext }>
						<Story />
					</ViewContextProvider>
				</WithRegistrySetup>
			);
		},
	],
};
