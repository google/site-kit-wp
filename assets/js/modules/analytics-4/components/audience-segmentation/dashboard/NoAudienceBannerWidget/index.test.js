/**
 * NoAudienceBannerWidget component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { useIntersection as mockUseIntersection } from 'react-use';

/**
 * Internal dependencies
 */
import NoAudienceBannerWidget from '.';
import { fireEvent, render } from '../../../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	muteFetch,
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	waitForDefaultTimeouts,
} from '../../../../../../../../tests/js/utils';
import {
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
} from '../../../../../../googlesitekit/constants';
import { CORE_UI } from '../../../../../../googlesitekit/datastore/ui/constants';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../../../datastore/constants';
import { withWidgetComponentProps } from '../../../../../../googlesitekit/widgets/util';
import { availableAudiences } from '../../../../datastore/__fixtures__';
import * as tracking from '../../../../../../util/tracking';
import { AUDIENCE_SELECTION_PANEL_OPENED_KEY } from '../AudienceSelectionPanel/constants';
import { mockLocation } from '../../../../../../../../tests/js/mock-browser-utils';

jest.mock( 'react-use', () => ( {
	...jest.requireActual( 'react-use' ),
	useIntersection: jest.fn(),
} ) );

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'NoAudienceBannerWidget', () => {
	mockLocation();

	let registry;

	const WidgetWithComponentProps = withWidgetComponentProps(
		'analyticsNoAudienceBanner'
	)( NoAudienceBannerWidget );

	const audienceSettingsRegExp = new RegExp(
		'^/google-site-kit/v1/core/user/data/audience-settings'
	);

	beforeEach( () => {
		mockUseIntersection.mockImplementation( () => ( {
			isIntersecting: false,
			intersectionRatio: 0,
		} ) );

		registry = createTestRegistry();
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );
		provideModuleRegistrations( registry );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'should not render when configuredAudiences is not loaded', () => {
		muteFetch( audienceSettingsRegExp );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setAvailableAudiences( availableAudiences );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should not render when configuredAudiences is null (not set)', () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setAvailableAudiences( availableAudiences );

		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: null,
			isAudienceSegmentationWidgetHidden: false,
		} );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should not render when configured audience is matching available audiences', () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setAvailableAudiences( availableAudiences );

		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [ 'properties/12345/audiences/1' ],
			isAudienceSegmentationWidgetHidden: false,
		} );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should not render when all configured audiences are matching available audiences', () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setAvailableAudiences( availableAudiences );

		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [
				'properties/12345/audiences/1',
				'properties/12345/audiences/3',
			],
			isAudienceSegmentationWidgetHidden: false,
		} );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should not render when some configured audiences are matching available audiences', () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setAvailableAudiences( availableAudiences );

		registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
			configuredAudiences: [
				'properties/12345/audiences/1',
				'properties/12345/audiences/9',
			],
			isAudienceSegmentationWidgetHidden: false,
		} );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	describe( "with an authenticated user who's never populated their audience selection", () => {
		let container, getByRole, getByText, rerender;

		beforeEach( () => {
			provideSiteInfo( registry );

			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setAvailableAudiences( availableAudiences );

			registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
				configuredAudiences: [],
				didSetAudiences: false,
			} );

			( { container, getByRole, getByText, rerender } = render(
				<WidgetWithComponentProps />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
				}
			) );
		} );

		it( 'should render correctly.', () => {
			expect(
				getByText( /You don’t have any visitor groups selected./i )
			).toBeInTheDocument();

			expect( container ).toMatchSnapshot();
		} );

		it( 'should open the Audience Selection Panel when clicking on "Select groups"', async () => {
			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBeUndefined();

			fireEvent.click( getByRole( 'button', { name: 'Select groups' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBe( true );
		} );

		it( 'should redirect to the settings page when clicking on "Settings"', async () => {
			expect( global.location.assign ).not.toHaveBeenCalled();

			fireEvent.click( getByRole( 'button', { name: 'Settings' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( global.location.assign ).toHaveBeenCalledWith(
				'http://example.com/wp-admin/admin.php?page=googlesitekit-settings#/admin-settings'
			);
		} );

		it( 'should track an event when the banner is viewed', () => {
			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps /> );

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-no-audiences',
				'view_banner',
				'none-selected'
			);
		} );

		it( 'should track an event when clicking on "Select groups"', async () => {
			fireEvent.click( getByRole( 'button', { name: 'Select groups' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-no-audiences',
				'select_groups',
				'none-selected'
			);
		} );

		it( 'should track an event when clicking on "Settings"', async () => {
			fireEvent.click( getByRole( 'button', { name: 'Settings' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-no-audiences',
				'change_settings',
				'none-selected'
			);
		} );
	} );

	describe( "with an authenticated user who's previously populated their audience selection", () => {
		let container, getByRole, getByText, rerender;

		beforeEach( () => {
			provideSiteInfo( registry );

			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setAvailableAudiences( availableAudiences );

			registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
				configuredAudiences: [],
				didSetAudiences: true,
			} );

			( { container, getByRole, getByText, rerender } = render(
				<WidgetWithComponentProps />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
				}
			) );
		} );

		it( 'should render correctly.', () => {
			expect(
				getByText(
					/It looks like your visitor groups aren’t available anymore./i
				)
			).toBeInTheDocument();

			expect( container ).toMatchSnapshot();
		} );

		it( 'should open the Audience Selection Panel when clicking on "Select other groups"', async () => {
			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBeUndefined();

			fireEvent.click(
				getByRole( 'button', { name: 'Select other groups' } )
			);

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBe( true );
		} );

		it( 'should redirect to the settings page when clicking on "Settings"', async () => {
			expect( global.location.assign ).not.toHaveBeenCalled();

			fireEvent.click( getByRole( 'button', { name: 'Settings' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( global.location.assign ).toHaveBeenCalledWith(
				'http://example.com/wp-admin/admin.php?page=googlesitekit-settings#/admin-settings'
			);
		} );

		it( 'should track an event when the banner is viewed', () => {
			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps /> );

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-no-audiences',
				'view_banner',
				'no-longer-available'
			);
		} );

		it( 'should track an event when clicking on "Select other groups"', async () => {
			fireEvent.click(
				getByRole( 'button', { name: 'Select other groups' } )
			);

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-no-audiences',
				'select_groups',
				'no-longer-available'
			);
		} );

		it( 'should track an event when clicking on "Settings"', async () => {
			fireEvent.click( getByRole( 'button', { name: 'Settings' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboard_audiences-no-audiences',
				'change_settings',
				'no-longer-available'
			);
		} );
	} );

	describe( "with a view-only user who's never populated their audience selection", () => {
		let container, getByRole, getByText, rerender;

		beforeEach( () => {
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setAvailableAudiences( availableAudiences );

			registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
				configuredAudiences: [],
				didSetAudiences: false,
			} );

			( { container, getByRole, getByText, rerender } = render(
				<WidgetWithComponentProps />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
				}
			) );
		} );

		it( 'should render correctly.', () => {
			expect(
				getByText( /You don’t have any visitor groups selected./i )
			).toBeInTheDocument();

			expect( container ).toMatchSnapshot();
		} );

		it( 'should open the Audience Selection Panel when clicking on "Select groups"', async () => {
			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBeUndefined();

			fireEvent.click( getByRole( 'button', { name: 'Select groups' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBe( true );
		} );

		it( 'should track an event when the banner is viewed', () => {
			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps /> );

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboardViewOnly_audiences-no-audiences',
				'view_banner',
				'none-selected'
			);
		} );

		it( 'should track an event when clicking on "Select groups"', async () => {
			fireEvent.click( getByRole( 'button', { name: 'Select groups' } ) );

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboardViewOnly_audiences-no-audiences',
				'select_groups',
				'none-selected'
			);
		} );
	} );

	describe( "with a view-only user who's previously populated their audience selection", () => {
		let container, getByRole, getByText, rerender;

		beforeEach( () => {
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.setAvailableAudiences( availableAudiences );

			registry.dispatch( CORE_USER ).receiveGetAudienceSettings( {
				configuredAudiences: [],
				didSetAudiences: true,
			} );

			( { container, getByRole, getByText, rerender } = render(
				<WidgetWithComponentProps />,
				{
					registry,
					viewContext: VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
				}
			) );
		} );

		it( 'should render correctly.', () => {
			expect(
				getByText(
					/It looks like your visitor groups aren’t available anymore./i
				)
			).toBeInTheDocument();

			expect( container ).toMatchSnapshot();
		} );

		it( 'should open the Audience Selection Panel when clicking on "Select other groups"', async () => {
			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBeUndefined();

			fireEvent.click(
				getByRole( 'button', { name: 'Select other groups' } )
			);

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect(
				registry
					.select( CORE_UI )
					.getValue( AUDIENCE_SELECTION_PANEL_OPENED_KEY )
			).toBe( true );
		} );

		it( 'should track an event when the banner is viewed', () => {
			expect( mockTrackEvent ).toHaveBeenCalledTimes( 0 );

			// Simulate the CTA becoming visible.
			mockUseIntersection.mockImplementation( () => ( {
				isIntersecting: true,
				intersectionRatio: 1,
			} ) );

			rerender( <WidgetWithComponentProps /> );

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboardViewOnly_audiences-no-audiences',
				'view_banner',
				'no-longer-available'
			);
		} );

		it( 'should track an event when clicking on "Select other groups"', async () => {
			fireEvent.click(
				getByRole( 'button', { name: 'Select other groups' } )
			);

			// Allow the `trackEvent()` promise to resolve.
			await waitForDefaultTimeouts();

			expect( mockTrackEvent ).toHaveBeenCalledTimes( 1 );
			expect( mockTrackEvent ).toHaveBeenCalledWith(
				'mainDashboardViewOnly_audiences-no-audiences',
				'select_groups',
				'no-longer-available'
			);
		} );
	} );
} );
