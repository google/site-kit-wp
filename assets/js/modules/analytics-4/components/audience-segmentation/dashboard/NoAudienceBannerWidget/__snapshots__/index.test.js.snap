// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoAudienceBannerWidget with a view-only user who's never populated their audience selection should render correctly. 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--analyticsNoAudienceBanner googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-lean-cta-banner googlesitekit-no-audience-banner"
      >
        <div
          class="googlesitekit-lean-cta-banner__body"
        >
          <div
            class="googlesitekit-lean-cta-banner__body-icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-lean-cta-banner__body-content"
          >
            <p>
              You don’t have any visitor groups selected. 
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Select groups
                </span>
              </button>
              .
            </p>
          </div>
        </div>
        <div
          class="googlesitekit-lean-cta-banner__graphic"
        >
          <svg />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NoAudienceBannerWidget with a view-only user who's previously populated their audience selection should render correctly. 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--analyticsNoAudienceBanner googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-lean-cta-banner googlesitekit-no-audience-banner"
      >
        <div
          class="googlesitekit-lean-cta-banner__body"
        >
          <div
            class="googlesitekit-lean-cta-banner__body-icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-lean-cta-banner__body-content"
          >
            <p>
              It looks like your visitor groups aren’t available anymore. 
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Select other groups
                </span>
              </button>
              .
            </p>
          </div>
        </div>
        <div
          class="googlesitekit-lean-cta-banner__graphic"
        >
          <svg />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NoAudienceBannerWidget with an authenticated user who's never populated their audience selection should render correctly. 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--analyticsNoAudienceBanner googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-lean-cta-banner googlesitekit-no-audience-banner"
      >
        <div
          class="googlesitekit-lean-cta-banner__body"
        >
          <div
            class="googlesitekit-lean-cta-banner__body-icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-lean-cta-banner__body-content"
          >
            <p>
              You don’t have any visitor groups selected. 
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Select groups
                </span>
              </button>
              .
            </p>
            <p>
              You can deactivate this widget in 
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Settings
                </span>
              </button>
              .
            </p>
          </div>
        </div>
        <div
          class="googlesitekit-lean-cta-banner__graphic"
        >
          <svg />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NoAudienceBannerWidget with an authenticated user who's previously populated their audience selection should render correctly. 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--analyticsNoAudienceBanner googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-lean-cta-banner googlesitekit-no-audience-banner"
      >
        <div
          class="googlesitekit-lean-cta-banner__body"
        >
          <div
            class="googlesitekit-lean-cta-banner__body-icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-lean-cta-banner__body-content"
          >
            <p>
              It looks like your visitor groups aren’t available anymore. 
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Select other groups
                </span>
              </button>
              .
            </p>
            <p>
              You can deactivate this widget in 
              <button
                class="googlesitekit-cta-link googlesitekit-cta-link--secondary"
              >
                <span
                  class="googlesitekit-cta-link__contents"
                >
                  Settings
                </span>
              </button>
              .
            </p>
          </div>
        </div>
        <div
          class="googlesitekit-lean-cta-banner__graphic"
        >
          <svg />
        </div>
      </div>
    </div>
  </div>
</div>
`;
